File: fft_rust/src/conv_naive.rs
// 参考: https://www.geeksforgeeks.org/program-to-generate-an-array-having-convolution-of-two-given-arrays/
use std::default::Default;
use std::ops::{Add, AddAssign, Div, Mul};

// 1の実装の方がシンプル

// time: O(N^2)
fn convolution_1<T>(array_1: &[T], array_2: &[T]) -> Vec<T>
where
    T: Default + Copy + Add + Mul + Div + AddAssign<<T as Mul>::Output>,
{
    let (len_1, len_2) = (array_1.len(), array_2.len());
    let mut result = vec![Default::default(); len_1 + len_2 - 1];

    for i in 0..len_1 {
        for j in 0..len_2 {
            result[i + j] += array_1[i] * array_2[j];
        }
    }

    result
}

// time: O(N^2)
fn convolution_2<T>(array_1: &[T], array_2: &[T]) -> Vec<T>
where
    T: Default + Copy + Add + Mul + Div + AddAssign<<T as Mul>::Output>,
{
    let (len_1, len_2) = (array_1.len(), array_2.len());
    let mut result = vec![Default::default(); len_1 + len_2 - 1];

    for k in 0..len_1 + len_2 - 1 {
        for i in 0..len_1 {
            if i <= k && k - i < len_2 {
                result[k] += array_1[i] * array_2[k - i];
            }
        }
    }

    result
}

fn main() {
    let case_1 = ([1, 2, 3, 4], [5, 6, 7, 8, 9]);
    println!("res: conv_1: {:?}", convolution_1(&case_1.0, &case_1.1));
    println!("res: conv_2: {:?}", convolution_2(&case_1.0, &case_1.1));
}

File: fft_cpp/conv_naive.cpp
#include <iostream>
#include <vector>

// time: O(N^2)
template<class T>
std::vector<T> convolution_1(const std::vector<T>& array_1, const std::vector<T>& array_2) {
    auto [len_1, len_2] = std::pair(array_1.size(), array_2.size());
    std::vector<T> result(len_1 + len_2 - 1);

    for(std::size_t i = 0; i < len_1; ++i) {
        for(std::size_t j = 0; j < len_2; ++j) {
            result[i + j] += array_1[i] * array_2[j];
        }
    }

    return result;
}

template<class T>
void show_result(const std::vector<T>& result) {
    for(auto& v: result) {
        std::cout << v << " ";
    }
    std::cout << std::endl;
}

int main(void) {
    auto case_1 = std::vector{1, 2, 3, 4};
    auto case_2 = std::vector{5, 6, 7, 8};
    auto res_1 = convolution_1(case_1, case_2);

    show_result(res_1);
}
